### PR TITLE
Icinga DB runtime updates and state via streams

### PR DESCRIPTION
**Pull request by N-o-X**
This PR changes Icinga DB runtime updates to use streams instead of pub/sub and changes the state sync to use runtime updates. This removes some race conditions when working with runtime updates in Icinga DB and also cleans up the process a lot.

### CODE REVIEW

File: lib/icingadb/icingadb-objects.cpp [hunk 1]
```diff
@@ -1037,15 +1048,24 @@ void IcingaDB::SendConfigUpdate(const ConfigObject::Ptr& object, bool runtimeUpd
 
 	String typeName = GetLowerCaseTypeNameDB(object);
 
-	std::map<String, std::vector<String>> hMSets, publishes;
-	std::vector<String> states 							= {"HMSET", m_PrefixStateObject + typeName};
+	std::map<String, std::vector<String>> hMSets;
+	std::vector<Dictionary::Ptr> runtimeUpdates;
+	std::vector<String> states = {"HMSET", m_PrefixConfigObject + typeName + ":state"};
```
Comment (julianbrost) [inline lib/icingadb/icingadb-objects.cpp:1053]: `states` seems to be an unused variable.

File: lib/icingadb/icingadb-objects.cpp [hunk 2]
```diff
@@ -1388,20 +1420,24 @@ void IcingaDB::SendConfigDelete(const ConfigObject::Ptr& object)
...
 
 	m_Rcon->FireAndForgetQueries({
-								   {"HDEL",    m_PrefixConfigObject + typeName, objectKey},
-								   {"DEL",     m_PrefixStateObject + typeName + ":" + objectKey},
-								   {"PUBLISH", "icinga:config:delete:" + typeName, objectKey}
+								   {"HDEL", m_PrefixConfigObject + typeName, objectKey},
+								   {
+								   		"XADD", "icinga:runtime:delete", "MAXLEN", "~", "1000000", "*",
+										"redis_key", m_PrefixConfigObject + typeName, "id", objectKey
+								   }
```
Comment (julianbrost) [inline lib/icingadb/icingadb-objects.cpp:1427]: Weird-looking indentation.

File: lib/icingadb/icingadb-objects.cpp [hunk 3]
```diff
@@ -2037,7 +2075,10 @@ Dictionary::Ptr IcingaDB::SerializeState(const Checkable::Ptr& checkable)
 
 	tie(host, service) = GetHostService(checkable);
 
-	attrs->Set("id", GetObjectIdentifier(checkable));;
+	String id = GetObjectIdentifier(checkable);
+
+	attrs->Set("id", id);
+	attrs->Set("object_id", id);
```
Comment (julianbrost) [inline lib/icingadb/icingadb-objects.cpp:2081]: Why is this added with two different keys?

**Review by julianbrost — CHANGES_REQUESTED** [review]
1. Why do the XADDs use MAXLEN? Wouldn't this just lead to silently lost updates if Icinga DB can't keep up?

2. The following seems to be a repeating pattern of writing into an element of `hMSets` and then, if it's a runtime update, perform a corresponding call to `AddObjectDataToRuntimeUpdates()` with the same redis key and data. I haven't checked each instance in detail but would it make sense to have a helper function for this?

**Review by N-o-X — COMMENTED** [review]
1. That's basically our cleanup. If Icinga DB can't keep up and we have over 1 million updates in the queue, we have bigger problems then that.
2. Yes, we could think about cleaning that up a bit. But that should not be part of this PR, I think.

File: lib/icingadb/icingadb-objects.cpp (hunk 3 above)
Comment (julianbrost) [reply lib/icingadb/icingadb-objects.cpp:2080]: To rephrase it (what helped me to understand it):
* `id`: This is the ID of the host/service state object
* `object_id`: This is the foreign key to the object (i.e. host for host states and service for service states) that this object is the state of

As there is a 1:1 relationship between host and host state, the host ID is also used as the host state ID. These are duplicated to 1) avoid having special handling for this in Icinga DB and 2) to have both a primary key and a foreign key in the SQL database in the end.

I think we should add this as a comment. It took me quite some time (and additional discussion) to understand this from the code, even though we talked about this before.

**Review by julianbrost — CHANGES_REQUESTED** [review]

File: lib/icingadb/icingadb-objects.cpp (hunk 3 above)
Comment (julianbrost) [reply lib/icingadb/icingadb-objects.cpp:2086]: Now there's a mix of "host" and "checkable" in the comment. I'd avoid using "checkable" in the comment and just describe it for the host and then end with "This applies to services analogously."

Also I'd rephrase it to better show what happens with the values: "[...], the host ID ('object_id', ends up as a foreign key 'host_state.host_id' referring to 'host.id' in the database) is also used as the host state ID ('id', ends up a the primary key 'host_state.id')"

**Review by julianbrost — CHANGES_REQUESTED** [review]

**Comment by julianbrost** [comment]
I have a general concern about having separate streams for insert/update (`icinga:runtime:upsert`) and delete (`icinga:runtime:delete`).

Consider a scenario where an object deletion follows shortly after an update of the same object. I think it would be possible for the message in `icinga:runtime:delete` to overtake the one in `icinga:runtime:upsert`, so Icinga DB could first delete the object and then recreate it.

**Comment by lippserd** [comment]
> I have a general concern about having separate streams for insert/update (`icinga:runtime:upsert`) and delete (`icinga:runtime:delete`).
> 
> Consider a scenario where an object deletion follows shortly after an update of the same object. I think it would be possible for the message in `icinga:runtime:delete` to overtake the one in `icinga:runtime:upsert`, so Icinga DB could first delete the object and then recreate it.

We could combine these streams as we already have an "if" for the key on the Icinga DB side. Another `if` for update/delete, won't hurt more. However, the race of updates  after delete is not resolved as Icinga DB handles this in parallel.
Icinga DB requires a lock so that only one operation per table can be active. A stream would still help here because we don't have to sort updates and deletions by time as well.

File: lib/icingadb/icingadb-objects.cpp [hunk 4]
```diff
@@ -242,12 +245,20 @@ void IcingaDB::UpdateAllConfigObjects()
...
+				std::vector<Dictionary::Ptr> runtimeUpdates;
+				CreateConfigUpdate(object, lcType, hMSets, runtimeUpdates, false);
 
 				// Write out inital state for checkables
 				if (dumpState) {
-					states.emplace_back(GetObjectIdentifier(object));
-					states.emplace_back(JsonEncode(SerializeState(dynamic_pointer_cast<Checkable>(object))));
+					String objectKey = GetObjectIdentifier(object);
+					Dictionary::Ptr state = SerializeState(dynamic_pointer_cast<Checkable>(object));
+					String checksum = HashValue(state);
```
Comment (julianbrost) [inline lib/icingadb/icingadb-objects.cpp:255]: So in the end, this hash will decide when Icinga DB will update `host_state`/`service_state` in MySQL, right?

Comment (N-o-X) [reply lib/icingadb/icingadb-objects.cpp:255]: Correct, it's used during the initial sync got check for changes.

**Review by julianbrost — APPROVED** [review]

File: lib/icingadb/icingadb-objects.cpp (hunk 3 above)
Comment (Al2Klimov) [reply lib/icingadb/icingadb-objects.cpp:2126]: Can’t Icinga DB grab object_id from id?

File: lib/icingadb/icingadb-objects.cpp [hunk 5]
```diff
@@ -631,18 +617,21 @@ void IcingaDB::InsertObjectDependencies(const ConfigObject::Ptr& object, const S
 				if (runtimeUpdate || m_DumpedGlobals.CustomVar.IsNew(kv.first)) {
 					allCvs.emplace_back(kv.first);
 					allCvs.emplace_back(JsonEncode(kv.second));
-				}
 
-				if (runtimeUpdate) {
-					publishes["icinga:config:update:customvar"].emplace_back(kv.first);
+					if (runtimeUpdate) {
+						Dictionary::Ptr data = kv.second;
```
[the remark this answers is not in the document]
Comment (N-o-X) [reply lib/icingadb/icingadb-objects.cpp:622]: I was unable to do an inline cast to Dictionary within the function call.

File: lib/icingadb/icingadb-objects.cpp [hunk 6]
```diff
@@ -926,13 +925,16 @@ void IcingaDB::InsertObjectDependencies(const ConfigObject::Ptr& object, const S
...
 
 			notificationRecipients.emplace_back(id);
-			notificationRecipients.emplace_back(JsonEncode(new Dictionary({{"notification_id", objectKey}, {"environment_id", m_EnvironmentId}, {"usergroup_id", usergroupId}})));
+			Dictionary::Ptr notificationRecipientData = new Dictionary({{"notification_id", objectKey}, {"environment_id", m_EnvironmentId}, {"usergroup_id", usergroupId}});
+			notificationRecipients.emplace_back(JsonEncode(notificationRecipientData));
 
 			if (runtimeUpdate) {
-				publishes["icinga:config:update:" + typeName + ":usergroup"].emplace_back(id);
+				AddObjectDataToRuntimeUpdates(runtimeUpdates, id, m_PrefixConfigObject + typeName + ":usergroup", groupData);
+				AddObjectDataToRuntimeUpdates(runtimeUpdates, id, m_PrefixConfigObject + typeName + ":recipient", notificationRecipientData);
```
[the remark this answers is not in the document]
Comment (N-o-X) [reply lib/icingadb/icingadb-objects.cpp:936]: Yes? We have `usergroup` and `recipient`? Maybe I don't understand your question?

File: lib/icingadb/icingadb-utility.cpp [hunk 7]
```diff
@@ -212,3 +212,19 @@ String IcingaDB::GetLowerCaseTypeNameDB(const ConfigObject::Ptr& obj)
...
+{
+	switch (value.GetType()) {
+		case ValueNumber:
+			return Convert::ToString(value);
+		case ValueBoolean:
+			return Convert::ToString(int(value));
+		case ValueString:
+			return Utility::ValidateUTF8(value);
+		case ValueEmpty:
+			return Convert::ToString(value);
```
[the remark this answers is not in the document]
Comment (N-o-X) [reply lib/icingadb/icingadb-utility.cpp:225]: I combined it.

File: lib/icingadb/icingadb-objects.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Al2Klimov) [reply lib/icingadb/icingadb-objects.cpp:621]: Have you tried making the parameter a `const Ptr&`?

File: lib/icingadb/icingadb-objects.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Al2Klimov) [reply lib/icingadb/icingadb-objects.cpp:936]: And the recipients are already handled below, aren’t they?

[the remark this answers is not in the document]
Comment (N-o-X) [reply lib/icingadb/icingadb-objects.cpp:936]: Nope, this is handling user groups and the part below handles the users.

File: lib/icingadb/icingadb-objects.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (N-o-X) [reply lib/icingadb/icingadb-objects.cpp:621]: But then I would have to cast back to a Dictionary to access and modify the underlaying data, right? It seems way easier to just have this extra line here.

[the remark this answers is not in the document]
Comment (Al2Klimov) [reply lib/icingadb/icingadb-objects.cpp:621]: Try it out. If an extra cast is required, the compiler will fail.

[the remark this answers is not in the document]
Comment (N-o-X) [reply lib/icingadb/icingadb-objects.cpp:621]: I really don't think that it's a good idea to use a more generic type, just because we want to avoid one extra line. What's the problem with this line?

[the remark this answers is not in the document]
Comment (Al2Klimov) [reply lib/icingadb/icingadb-objects.cpp:621]: Which "more generic type"? Have you tried to change the function parameter's type `Dictionary::Ptr&` -> `const Dictionary::Ptr&` or not?

[the remark this answers is not in the document]
Comment (N-o-X) [reply lib/icingadb/icingadb-objects.cpp:621]: Ohh, I totally misunderstood what you meant. That seems to work 👍
